### DISCUSSION TITLE
Recommended WIndows font or terminal?

### DISCUSSION THREAD

**Discussion by landonjx** [Textualize/textual#3117] [Q&A] [answered]
Are there any recommended fonts that work well with Textual on Windows?

Should
* Include all of the graphical and graphical-like characters Textual might use
* Include the "normal" set of characters
* Be sans serif
* Be free (as in beer)

Windows' Command and PowerShell terminals do not have a separate line height option (like Mac) and most of the pre-installed monospace fonts do not include all the graphical characters used by Textual. For example, Lucida Console and Consolas do not include any of the graphical characters, many of the fonts that do include some graphical characters, or do not include the circle (used at the top left corner of the demo). The only built-in font that have all the symbols seems to be MS Gothic (but appears to have remapped `\`, maybe I need to use a different code page with this one?) and SimSun (which is a serifs fonts).

I've tried a number of fonts over on Google Fonts but so far they do not included the graphical characters.

**Comment by TomJGooding** [comment]
I'm not sure if pre-installed on Windows, but try `DejaVu Sans Mono` which should have decent Unicode support.

The font used in the Textual docs looks to be `Fira Code` which might be worth a try.

**Comment by TomJGooding** [comment]
I've seen a few issues related to fonts, so might be worth adding an 'official' answer to the FAQ and/or docs?

**Comment by ouroborus** [comment]
Fira Code defaults to a larger line height than it needs so has artifacts on some of the vertical line draw characters. It's also missing the circle.

DejaVu Sans Mono looks really nice, except it's also missing the circle character.

(I'm not sure which codepoint the circle is and it only seems to be used in the "main" example (`python -m textual`) so I'm not sure which widget is using it. Maybe it's not a big deal?)

**Comment by TomJGooding** [comment]
The circle glpyh is used as the icon in the [Header](https://github.com/Textualize/textual/blob/main/src/textual/widgets/_header.py ) widget. Not sure if it might also be used anywhere else.

Forgive my ignorance of Windows terminals, but do they not use a fallback font?

**Comment by ouroborus** [comment]
They don't. If a font doesn't have a character, you get whatever the "missing character" character is, usually a box, box with question mark, or sometimes just a question mark.

**Comment by TomJGooding** [comment]
You could try installing [Windows Terminal](https://apps.microsoft.com/store/detail/windows-terminal/9N0DX20HK701), which looks to have font fallback and generally seems to display Textual apps better. 

Or look at a patched font such as those from the [Nerd Fonts project](https://www.nerdfonts.com/).

**Comment by willmcgugan** [comment]
I've only ever used Windows Terminal on Windows 11 with default settings. It uses Cascadia Mono, and I don't see any glitched characters.

**Comment by TomJGooding** [comment]
@willmcgugan Cascadia Mono doesn't include the circle glyph (U+2B58) either.  But Windows Terminal seems to have font fallback for unsupported glyphs, unlike the classic Windows conhost.

Font fallback can also have its drawbacks, depending on the terminal/environment. See this discussion #3091 last week for example.

I've seen a few issues related to fonts, so might be worth adding an 'official' answer to the FAQ and/or docs?

**Comment by ouroborus** [comment, marked as answer]
I (and @landonjx) are on Win10. I was able to find a Win11 system to try out @willmcgugan's assertion. It does indeed look great and all the characters I was able to test are rendered correctly.

From summarizing and testing:

**Cascadia Mono** works well on both Win10 and Win11. It's missing the circle glyph but the font fallback available in **Windows Terminal** (installed by default in Win11, [available in the Microsoft Store](https://apps.microsoft.com/store/detail/windows-terminal/9N0DX20HK701) otherwise, (Thanks @willmcgugan & @TomJGooding !) ) fixes that. If you don't already have the font, it can be found in the [releases](https://github.com/microsoft/cascadia-code/releases) in the [Cascadia Code repo](https://github.com/microsoft/cascadia-code). (The repo includes both but note that Cascadia Code has issues when used with Textual, at least in Win10.)

On Win10's default consoles, [DevaVuSansM Nerd Font](https://www.nerdfonts.com/font-downloads) works well. (Thanks @TomJGooding !) It seems to include all the needed symbols, including the circle glyph, and there are absolutely no artifacts.

**Comment by davep** [comment]
I'm going to move this issue over to a discussion, under Q&A. It looks handy enough to keep around in a more A&Q capacity.
